File: rayt-python/rayt_python/sphere.py
import typing

from rayt_python.hittable import HitRecord, Hittable
from rayt_python.material import Material
from rayt_python.ray import Ray
from rayt_python.vec3 import Point3, dot


class Sphere(Hittable):
    def __init__(self, cen: Point3, r: float, m: Material) -> None:
        self.center = cen
        self.radius = r
        self.material = m

    def create_rec(self, ray: Ray, t: float) -> HitRecord:
        p = ray.at(t)
        rec = HitRecord(p=p, t=t, material=self.material)
        rec.set_face_normal(ray, (p - self.center) / self.radius)
        return rec

    def hit(self, r: Ray, t_min: float, t_max: float) -> typing.Union[HitRecord, None]:
        oc = r.origin - self.center
        a = r.direction.length_squared
        half_b = dot(oc, r.direction)
        c = oc.length_squared - pow(self.radius, 2)
        discriminant = pow(half_b, 2) - a * c

        if discriminant > 0:
            root = pow(discriminant, 0.5)
            temp = (-half_b - root) / a
            if temp < t_max and temp > t_min:
                return self.create_rec(r, temp)

            temp = (-half_b + root) / a
            if t_min < temp < t_max:
                return self.create_rec(r, temp)

        return None

File: rayt-rust/src/sphere.rs
use crate::{
    hittable::{HitRecord, Hittable},
    material::Material,
    ray::Ray,
    vec3::{dot, Point3},
};

pub(crate) struct Sphere {
    center: Point3,
    radius: f64,
    material: Box<dyn Material>,
}

impl Sphere {
    pub(crate) fn new(center: Point3, radius: f64, material: Box<dyn Material>) -> Self {
        Self {
            center,
            radius,
            material,
        }
    }

    fn create_rec(self, r: &Ray, t: f64) -> HitRecord {
        let p = r.at(t);
        let mut rec = HitRecord::new(&p, t, self.material);
        rec.set_face_normal(r, &((p - self.center) / self.radius));
        rec
    }
}

impl Hittable for Sphere {
    fn hit(&self, r: &Ray, t_min: f64, t_max: f64) -> Option<HitRecord> {
        let oc = r.origin - self.center;
        let a = r.direction.length_squared();
        let half_b = dot(&oc, &r.direction);
        let c = oc.length_squared() - self.radius.powi(2);
        let discriminant = half_b.powi(2) - a * c;

        if discriminant > 0.0 {
            let root = f64::sqrt(discriminant);
            let temp = (-half_b - root) / a;
            if temp < t_max && temp > t_min {
                return Some(self.create_rec(r, temp));
            }

            temp = (-half_b + root) / a;
            if temp > t_min && temp < t_max {
                return Some(self.create_rec(r, temp));
            }
        }

        None
    }
}
